src/txn/txn_ckpt.c

```diff
@@ -15,6 +15,153 @@ static int __checkpoint_presync(WT_SESSION_IMPL *, const char *[]);
 
+/*
+ * __checkpoint_flush_tier_wait --
+ *     Wait for all previous work units queued to be processed.
+ */
+static int
+__checkpoint_flush_tier_wait(WT_SESSION_IMPL *session, const char **cfg)
+{
+    WT_CONFIG_ITEM cval;
+    WT_CONNECTION_IMPL *conn;
+    uint64_t now, start, timeout;
+    int yield_count;
+
+    conn = S2C(session);
+    yield_count = 0;
+    now = start = 0;
+    /*
+     * The internal thread needs the schema lock to perform its operations and flush tier also
+     * acquires the schema lock. We cannot be waiting in this function while holding that lock or no
+     * work will get done.
+     */
+    WT_ASSERT(session, !FLD_ISSET(session->lock_flags, WT_SESSION_LOCKED_SCHEMA));
+    WT_RET(__wt_config_gets(session, cfg, "flush_tier.timeout", &cval));
+    timeout = (uint64_t)cval.val;
+    if (timeout != 0)
+        __wt_seconds(session, &start);
+
+    /*
+     * It may be worthwhile looking at the add and decrement values and make choices of whether to
+     * yield or wait based on how much of the workload has been performed. Flushing operations could
+     * take a long time so yielding may not be effective.
+     */
+    while (!WT_FLUSH_STATE_DONE(conn->flush_state)) {
+        if (start != 0) {
+            __wt_seconds(session, &now);
+            if (now - start > timeout)
+                return (EBUSY);
+        }
+        if (++yield_count < WT_THOUSAND)
+            __wt_yield();
+        else {
+            __wt_cond_signal(session, conn->tiered_cond);
+            __wt_cond_wait(session, conn->flush_cond, 200, NULL);
+        }
+    }
+    return (0);
+}
+
+/*
+ * __checkpoint_flush_tier --
+ *     Perform one iteration of tiered storage maintenance.
+ */
+static int
+__checkpoint_flush_tier(WT_SESSION_IMPL *session, bool force)
+{
+    WT_CKPT ckpt;
+    WT_CONFIG_ITEM cval;
+    WT_CONNECTION_IMPL *conn;
+    WT_CURSOR *cursor;
+    WT_DECL_RET;
+    uint64_t ckpt_time, flush_time;
+    const char *key, *value;
+
+    __wt_verbose(session, WT_VERB_TIERED, "CKPT_FLUSH_TIER: Called force %d", force);
+
+    WT_STAT_CONN_INCR(session, flush_tier);
+    conn = S2C(session);
+    cursor = NULL;
+    /*
+     * For supporting splits and merge:
+     * - See if there is any merging work to do to prepare and create an object that is
+     *   suitable for placing onto tiered storage.
+     * - Do the work to create said objects.
+     * - Move the objects.
+     */
+    conn->flush_state = 0;
+
+    /*
+     * We hold the checkpoint lock so we know no other thread can be doing a checkpoint at this time
+     * but our time can move backward with respect to the time set by a different thread that did a
+     * checkpoint. Update time value for most recent flush_tier, taking the more recent of now or
+     * the checkpoint time.
+     */
+    WT_ASSERT(session, FLD_ISSET(session->lock_flags, WT_SESSION_LOCKED_CHECKPOINT));
+    __wt_seconds(session, &flush_time);
```

Comment:
At this point the checkpoint's time has already been set by `txn_checkpoint_establish_time`. So taking a new time here, we could wind up with a time that is greater than the checkpoint time. This will cause problems -- it could prevent a future flush from flushing tables that we modify in the checkpoint that's about to happen.

I think it will work to use the checkpoint time here, `session->current_ckpt_sec`. Below, we skip the switch & flush if a tables flush_time > checkpoint_time. Using the checkpoint time as the flush time would ensure that the next flush_tier will flush the checkpoint.

Reply:
Ok, what is the difference between `session->current_ckpt_sec` and `conn->ckpt_most_recent`? They seem really similar and it the `ckpt_most_recent` isn't, for example, set at the end from `current_ckpt_sec`?

Reply:
I think the significant difference is that `current_ckpt_sec` will stay the same even if someone else updates `ckpt_most_recent`. But I'm not sure who that "someone" would be. It looks like checkpoint is the only place we call `txn_checkpoint_establish_time` and there is no other code (except startup) that sets `ckpt_most_recent`.

Reply:
I think I fixed this. I used ckpt_most_recent.